### PR TITLE
Fix `IndexError` in `GitConfigParser` When a Quoted Config Value Contains a Trailing New Line

### PR DESCRIPTION
**Pull request by DaveLak**
Fixes: https://github.com/gitpython-developers/GitPython/issues/1887

Improve the guarding `if` check in `GitConfigParser`'s `string_decode` function to safely handle empty strings and prevent `IndexError`s when accessing string elements.

This resolves an IndexError in the `GitConfigParser`'s `.read()` method when the config file contains a quoted value ending with a new line.

### CODE REVIEW

File: fuzzing/fuzz-targets/fuzz_config.py
Comment (DaveLak) [inline fuzzing/fuzz-targets/fuzz_config.py:38]: The `isinstance` check is not strictly necessary, but I think it reads better with it, so I left it. I'm happy to remove it if that's preferred.

Comment (Byron) [reply fuzzing/fuzz-targets/fuzz_config.py:38]: When I read it, I wonder why it's necessary and can't figure it out.
And since it's not necessary, I think it would be better to remove the `isinstance` check.

Comment (Byron) [reply fuzzing/fuzz-targets/fuzz_config.py:38]:
```suggestion
        if "embedded null byte" in str(e):
```

**Review by Byron — APPROVED** [review]
Thanks a lot for this improvement!

I will merge right after CI is green after applying my tiny patch.

File: git/config.py
Comment (EliahKagan) [inline git/config.py:455]: I think this can be made a bit clearer (and more idiomatic) by doing:

```suggestion
            if v.endswith("\\"):
```

After all, we want to know if it ends with a backslash, which `endswith` safely and naturally expresses.

**Review by EliahKagan — COMMENTED** [review]
I've suggested a small change that I think improves clarity.

Separately from the changes here but relevant to the logic they are a part of, I'm a bit worried about the case where there are *two* (or more) trailing backslashes, so that removing one of them still leaves a trailing backslash. However, that was present before, and this PR definitely need not be expanded to address it! I'm also not certain I'm right to be worried about that; maybe it is less important to cover, or a part of a separate class of malformed configuration file inputs that are intended to trigger decoding errors.

File: git/config.py
Comment (Byron) [reply git/config.py:455]: I also think that's better, however, `v` seems to be nullable which isn't checked for anymore. I will apply the patch with that change, just to be sure.

Comment (Byron) [reply git/config.py:455]:
```suggestion
            if v and v.endswith("\\"):
```

**Comment by Byron** [comment]
> I've suggested a small change that I think improves clarity.
> 
> Separately from the changes here but relevant to the logic they are a part of, I'm a bit worried about the case where there are _two_ (or more) trailing backslashes, so that removing one of them still leaves a trailing backslash. However, that was present before, and this PR definitely need not be expanded to address it! I'm also not certain I'm right to be worried about that; maybe it is less important to cover, or a part of a separate class of malformed configuration file inputs that are intended to trigger decoding errors.

Thanks for taking another look - admittedly I forgot to merge this PR. Cygwin is the bane of this CI as it takes unnatural amounts of time :/.

Having worked on the `gitoxide` version of this I can say that this implementation here is more hole than cheese (if that makes sense :)), but it seems to work well enough for the common cases, still. Since GitPython is in maintenance mode, I don't think there is any need to try to improve it beyond fixing bugs.

File: git/config.py
Comment (EliahKagan) [inline git/config.py:455]: Is there a reason to separately check if `v` is truthy? In my mind, the reason to call `endswith` is that it takes care of the entire check in a single simple expression, regardless of whether the string was empty (falsy) or nonempty (truthy).

```suggestion
            if v.endswith("\\"):
```

Unless the parameter type annotation is incorrect and `v` could be some non-string such as the `None` object, I recommend against separately checking if `v` is truthy. The main benefit of calling `endswith` is that it does the full check regardless of whether the string is nonempty (truthy), making a more complex expression, such as one with the `and` operator, unnecessary.

The only way it would be necessary to check if `v` is truthy before calling `endswith` is if the parameter type annotation is incorrect and `v` is not always a string. If that is the case, then it would be a bigger issue (that's the main reason I'm posting this additional comment, even after the PR is merged).

Comment (EliahKagan) [reply git/config.py:455]: Sorry, for some reason I didn't see [that comment](https://github.com/gitpython-developers/GitPython/pull/1908#discussion_r1580488620)! That addresses this proactively and my above comment can be disregarded entirely.

If the code is modified further in the future then the `str` annotation, which is inconsistent with a `None` value, could be examined and potentially changed.

File: git/config.py
Comment (EliahKagan) [reply git/config.py]: For some reason, I didn't see, or carelessly glossed over, [that comment](https://github.com/gitpython-developers/GitPython/pull/1908#discussion_r1580488620), and this posted [this unnecessary comment](https://github.com/gitpython-developers/GitPython/pull/1908#discussion_r1580500339) that can be disregarded in its entirety. Unfortunately, GitHub doesn't allow me to mark that review thread as resolved, and I don't want to delete it lest that cause further confusion if it has been seen. My apologies to you both!

File: git/config.py
Comment (Byron) [reply git/config.py:455]: By the way, I tried to enable branch protections which should enable auto-merge, and thus resolve the issue with long-running CI sometimes preventing a merge in addition to me forgetting to do it later. It didn't yet work for this PR, but I will see for the next one.

**Pull request merged**
